File: docs/examples/push_client.py
"""Server push, dialer half: a dialer with an `on_stream` handler of its own.

Start `push_server.py` first, then run `python push_client.py`. The URL is `ws://127.0.0.1:8001/ws`
unless `MUXWS_URL` says otherwise.

The handler is passed to `connect()` rather than registered after it returns, because the acceptor
may push a stream the instant it sees this dialer: a handler registered one `await` later would meet
that push with `reset(REFUSED)`.
"""

import asyncio
import os

from muxws import connect, Stream

URL = os.environ.get("MUXWS_URL", "ws://127.0.0.1:8001/ws")


async def main() -> None:
    finished = asyncio.Event()

    async def on_push(payload: object, stream: Stream) -> None:
        print(f"push: {payload['topic']}")
        async for tick in stream:
            print(f"tick {tick['tick']}")
        finished.set()

    peer = await connect(URL, on_stream=on_push)
    try:
        # One-shot: `notify()` sends a payload and ends the stream, and hands back no handle to
        # await, because there is nothing coming back on it.
        await peer.notify({"subscribe": "ticks"})
        # 10.0 seconds, and only so a stuck example fails rather than hangs.
        await asyncio.wait_for(finished.wait(), timeout=10.0)
    finally:
        await peer.close()


if __name__ == "__main__":
    asyncio.run(main())

File: docs/examples/push-client.ts
/**
 * Server push, dialer half, in TypeScript on Node - the other end of `push_server.py`.
 *
 * Run it with `npx tsx push-client.ts`. The URL is `ws://127.0.0.1:8001/ws` unless `MUXWS_URL` says
 * otherwise.
 *
 * `onStream` is passed to `connect()` rather than registered after it returns, because the acceptor
 * may push a stream the instant it sees this dialer: a handler registered one `await` later would
 * meet that push with `reset(REFUSED)`.
 */

import 'muxws';
import { connect } from 'muxws/node';

const url = process.env.MUXWS_URL ?? 'ws://127.0.0.1:8001/ws';

interface Topic {
  topic: string;
}

interface Tick {
  tick: number;
}

let finished: () => void = () => undefined;
const pushDone = new Promise<void>((resolve) => {
  finished = resolve;
});

const peer = await connect(url, {
  onStream: async (payload, stream) => {
    console.log(`push: ${(payload as Topic).topic}`);
    for await (const tick of stream) {
      console.log(`tick ${(tick as Tick).tick}`);
    }
    finished();
  },
});

try {
  // One-shot: `notify()` sends a payload and ends the stream, and hands back no handle to await,
  // because there is nothing coming back on it.
  await peer.notify({ subscribe: 'ticks' });
  // 10000 milliseconds, and only so a stuck example fails rather than hangs.
  const deadline = new Promise<never>((_resolve, reject) => {
    setTimeout(() => {
      reject(new Error('the push never arrived within 10000 milliseconds'));
    }, 10_000).unref();
  });
  await Promise.race([pushDone, deadline]);
} finally {
  await peer.close();
}
